Implement adaptive hashing using specialization

pczarn opened the pull request:
Adaptive hashing provides fast and complexity-safe hashing for hashmaps with simple key types. The user doesn't need to change any code to get speedups from adaptive hashing, in contrast to the use of FnvHasher.

pczarn committed:
Implement adaptive hashing using specialization

pczarn committed:
Add benchmarks

apasel422 commented on src/lib.rs:
```diff
@@ -54,6 +59,8 @@ use table::BucketState::{
     Full,
 };
 
+pub use adaptive_map::HashMapInterface;
```
Does this need to be public?

pczarn replied:
Yes, this exists temporarily. As long as specialization doesn't work for inherent impls, users of adaptive hashing must import this trait.

apasel422 replied:
It seems like the inherent methods on `HashMap` could defer to `HashMapInterface`, no?

pczarn replied:
I'll try that, it's a great idea!

apasel422 commented on src/lib.rs:
```diff
@@ -394,12 +401,14 @@ fn pop_internal<K, V>(starting_bucket: FullBucketMut<K, V>) -> (K, V) {
 ///
 /// `hash`, `k`, and `v` are the elements to "robin hood" into the hashtable.
-fn robin_hood<'a, K: 'a, V: 'a>(mut bucket: FullBucketMut<'a, K, V>,
+pub fn robin_hood<'a, K: 'a, V: 'a, M: 'a>(mut bucket: FullBucket<K, V, M>,
```
Does this need to be public?

pczarn replied:
No, I'll make it private

pczarn commented:
Now, some benchmarks. With `black_box`, in `benches` directory. This benchmark indicates real differences the most:

#### Adaptive

```
Adaptive
find_existing     bench: 13,655 ns/iter (+/- 2,068)
find_nonexisting  bench: 15,241 ns/iter (+/- 411)
get_remove_insert bench:     68 ns/iter (+/- 8)
grow_by_insertion bench:    131 ns/iter (+/- 11)
hashmap_as_queue  bench:     60 ns/iter (+/- 8)
new_drop          bench:      3 ns/iter (+/- 0)
new_insert_drop   bench:     84 ns/iter (+/- 2)
```

#### (Outdated!! the 2-4 variant is no longer used) SipHash-2-4

```
find_existing     bench: 26,208 ns/iter (+/- 916)
find_nonexisting  bench: 24,833 ns/iter (+/- 5,082)
get_remove_insert bench:    118 ns/iter (+/- 0)
grow_by_insertion bench:    159 ns/iter (+/- 5)
hashmap_as_queue  bench:     82 ns/iter (+/- 0)
new_drop          bench:     76 ns/iter (+/- 3)
new_insert_drop   bench:    167 ns/iter (+/- 7)
```

apasel422 commented on src/lib.rs:
````diff
@@ -317,7 +323,7 @@ fn test_resize_policy() {
 /// ```
 #[derive(Clone)]
-pub struct HashMap<K, V, S = RandomState> {
+pub struct HashMap<K, V, S = AdaptiveState> {
````
Changing this would be a breaking change at the API level, right?

pczarn replied:
Yes, this change can't be added to the std library. Is this library a drop-in replacement for std's HashMap?

apasel422 replied:
I think the purpose of this repo is to iterate outside std, but with the ultimate goal of incorporating any worthwhile changes back into std.

pczarn commented:
Two problems remain.

~~First, adaptive maps are not yet safeguarded against DoS by insertion through **Entry**. The first step is tracking nightly (#6).~~

Second, the DoS safeguard ignores the effect of fn robin_hood. The safeguard is simple; consider a situation where the user does an insertion followed by searches. The insertion of an element ensures that searches for that element in the future won't take more than a limited number of iterations (a threshold of 128). However, in between element insertion and searches for that element, other unrelated insertions may displace that element.

It seems ~~clear~~ that fn robin_hood by definition can't increase the displacement of a "unfortunate" ousted element beyond the displacement of another more "fortunate" element. Since the latter is already below the threshold of 128, it follows that the former won't be increased above 128.

This should be closely reconsidered and documented in the form of a proof.

pczarn committed:
f update adaptive map for nightly

pczarn committed:
update benches

pczarn committed:
no private types in public interfaces; backwards compatibility

pczarn committed:
refactor a large function

pczarn committed:
doc

pczarn committed:
Use a trait instead of a macro for specialization

pczarn committed:
remove test

pczarn committed:
doc

pczarn committed:
Calculate probabilities

pczarn commented:
Updated and rebased.

/cc @contain-rs/publishers @bill-myers @pnkfelix @arthurprs @Jurily @gereeter @Ticki @divyekapoor
https://github.com/rust-lang/rfcs/issues/631, https://github.com/rust-lang/rust/issues/11783 "Implement fast HashMap"
https://github.com/rust-lang/rust/issues/29754 "Change Siphash to use one of the faster variants of the algorithm (Siphash13, Highwayhash)"
https://github.com/rust-lang/rust/pull/28044 "WIP: Hash and Hasher update for faster common case hashing"
Source of the idea: https://github.com/rust-lang/rust/issues/11783#issuecomment-34882467

arthurprs commented:
The difference is brutal.

Does this apply to &str/String as well?

pczarn commented:
> Does this apply to &str/String as well?

Not yet, because this requires explicit one-shot hashing with a hasher such as FarmHash, which probably needs an RFC.

ticki commented:
Wonderful, @pczarn. Great work!

Valloric commented:
Is this PR blocked on something? Why haven't there been any updates in 5 months?

pczarn commented:
The difference is now smaller, because SipHash-1-3 is used by default.

pczarn commented:
I'm currently writing an RFC.

arthurprs commented:
We should investigate using the integers as the hash in the Adaptive path.

pczarn commented:
@arthurprs I'm sure there's no good way of using integers as the hash. The adaptive path allows us to use a non-cryptographic hash, but the hash should still be statistically good.

However, we could use hashes that are cheaply converted back to integer keys. That is, a reversible function for hashing. That would save some space at the expense of slower access to integer keys. Unfortunately, HashMap exposes iterators that must borrow keys stored somewhere inside the structure. It's a dead end.

There's more to be gained from using 32-bit hashes on 32-bit platforms. Perhaps 48-bit hashes on 64-bit platforms, too.

arthurprs commented:
@pczarn I wouldn't be so sure. I'm almost positive that both g++ and clang uses the integer itself as the hash in the stdlib unordered_map (in these cases they don't even store the hash, but that's another optimization).

cgaebel commented:
Robin hood hashing is very slow with a statistically poor hash function.
Unordered_map is implemented differently than the rust hashmap.
On Wed, Jul 13, 2016 at 7:24 PM Arthur Silva notifications@github.com
wrote:

> @pczarn https://github.com/pczarn I wouldn't be so sure. I'm almost
> positive that both g++ and clang uses the integer itself as the hash in the
> stdlib unordered_map (in these cases they don't even store the hash, but
> that's another optimization).
> 
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/contain-rs/hashmap2/pull/5#issuecomment-232339210,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAH3yd8d3NY9IhWIHPcYmtztxbzdOTgFks5qVNj7gaJpZM4H2XyN
> .

arthurprs commented:
Well, I'm not an expert but it works very well in practice as it's a perfect hashing and moving collisions is most of the times cheap even compared to the finalizer computation. Initializing the hash with the thread local seed and xoring the scalar against it helps too.
In the case of hand crafted rogue input it'll fallback to sip, so no harm done (the same applies for any finalizer).

EDIT: I did this pretty quickly so I might have #$!!#$$ up something

My results for https://github.com/arthurprs/hashmap2/commit/03a4fb1cefa07c29e49a8bda92ce1d8d4ef6ced0

Adaptive Identity hash

```
test find_existing     ... bench:       7,265 ns/iter (+/- 1,251)
test find_nonexisting  ... bench:       7,184 ns/iter (+/- 614)
test get_remove_insert ... bench:          30 ns/iter (+/- 3)
test grow_by_insertion ... bench:          25 ns/iter (+/- 4)
test hashmap_as_queue  ... bench:          17 ns/iter (+/- 3)
test new_drop          ... bench:           1 ns/iter (+/- 0)
test new_insert_drop   ... bench:          53 ns/iter (+/- 19)
test rev_insert        ... bench:      12,189 ns/iter (+/- 1,034)
test rng_insert        ... bench:      16,854 ns/iter (+/- 1,457)
test seq_insert        ... bench:      12,148 ns/iter (+/- 904)
```

Adaptive Finalizer hash

```
test find_existing     ... bench:      29,724 ns/iter (+/- 2,711)
test find_nonexisting  ... bench:      30,017 ns/iter (+/- 2,539)
test get_remove_insert ... bench:         140 ns/iter (+/- 11)
test grow_by_insertion ... bench:         113 ns/iter (+/- 9)
test hashmap_as_queue  ... bench:          79 ns/iter (+/- 9)
test new_drop          ... bench:           1 ns/iter (+/- 0)
test new_insert_drop   ... bench:          71 ns/iter (+/- 2)
test rev_insert        ... bench:      36,587 ns/iter (+/- 3,314)
test rng_insert        ... bench:      32,512 ns/iter (+/- 6,206)
test seq_insert        ... bench:      33,651 ns/iter (+/- 1,276)
```

Sip24

```
test find_existing     ... bench:      28,713 ns/iter (+/- 2,338)
test find_nonexisting  ... bench:      29,154 ns/iter (+/- 4,531)
test get_remove_insert ... bench:         146 ns/iter (+/- 20)
test grow_by_insertion ... bench:         144 ns/iter (+/- 16)
test hashmap_as_queue  ... bench:          83 ns/iter (+/- 4)
test new_drop          ... bench:         122 ns/iter (+/- 10)
test new_insert_drop   ... bench:         189 ns/iter (+/- 10)
test rev_insert        ... bench:      12,674 ns/iter (+/- 1,652)
test rng_insert        ... bench:      16,246 ns/iter (+/- 5,278)
test seq_insert        ... bench:      12,902 ns/iter (+/- 1,573)
```

pczarn commented:
> In the case of hand crafted rogue input it'll fallback to sip

Not only with hand crafted input. With a poor hash function, normal input can be misidentified as an intentional attack. With an identity function, it only needs to contain some integers that differ by medium or large powers of two ± 10 or 20.

Falling back to sip is highly expensive. Adaptive hashing as implemented here tries very hard to never fall back. I would rather not give that up.

arthurprs commented:
It preserves divisibility so there's the obvious worst case (since the HashMap uses 2^N capacities). I'm positive that Robin Hood hashing really helps with this though.

Edit: considering we're xoring against the seed that wouldn't be a problem anymore?

The most\* popular languages in this _planet_ (Python, Java, C++) use the identity function for hashing integers, and they solve collisions in a weaker way too (except java probably). I don't think it's a coincidence that they accepted this tradeoff.

Unfortunately I can't come up with actual math to check if it doesn't interact badly with the adaptive code, so that can be the nail in the coffin for my argument.

divyekapoor commented:
Knowing how the hash function behaves and its strong and weak points allows
one to take decisions accordingly. Having no information at all means that
no reasoning can take place. It's not clear to me what guarantees the
adaptive hash provides, whether it holds up to uniformity of distribution
or has holes and under what circumstances the adaptations will progress.

It is not enough to state that the individual components of the adaptive
hash are well understood. How they interact with each other is crucial to
determine the overall expected distribution of hashed values. If we don't
have the math to nail this down, it's going to be very hard to justify this
as the default choice.

Looking forward to a formal analysis of this function and a
characterisation of its behaviour.

Cheers!
Divye
On Wed, Jul 13, 2016 at 11:18 AM Arthur Silva notifications@github.com
wrote:

> It preserves divisibility so there's the worst case is more obvious (since
> the HashMap uses 2^N capacities). I'm positive that Robin Hood hashing
> really helps with this though.
> 
> The most\* popular languages in this _planet_ (Python, Java, C++) use the
> identity function for hashing integers, and they solve collisions in a
> weaker way too (except java probably). I don't think it's a coincidence
> that they accepted this tradeoff.
> 
> Unfortunately I can't come up with actual math to check if doesn't
> interact badly with the adaptive code, so that can be the nail in the
> coffin for my argument.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/contain-rs/hashmap2/pull/5#issuecomment-232442218,
> or mute the thread
> https://github.com/notifications/unsubscribe/AANYcOojeUwCxpPb1b2AvUFTs0cdWUAHks5qVSwRgaJpZM4H2XyN
> .

pczarn commented:
@divyekapoor Adaptive Robin Hood is not a new hash function, it's a way of using two existing hash functions, h₁ and h₂, at different times. The latter is a cryptographic hash function, whereas the former is just a hash function with good statistical properties*. This implementation will use SipHash-1-3 as the latter, and Farmhash as the former.

Still, the name "adaptive hashing" is misleading. I'd like to use one of the following short names.
- Adaptive Robin Hood
- Fused Robin Hood hashing
- Safeguarded Robin Hood hashing
- Firewalled Robin Hood hashing
- Amortized Robin Hood hashing

\* All cryptographic hash functions have good avalanche effect, but not all hash functions that have good avalanche effect are cryptographic.

@arthurprs You're right about some languages using an identity function as h₁. However, HashMap in Rust has different needs, design choices and guarantees compared to many chaining hash maps. (Xoring against the seed doesn't help.)

Thanks for the benchmarks. Yes, they show oddly low speed for the finalizer hash. It does 2 multiplications and 6 basic operations, which should be very fast on most CPUs. (This is also an argument _against using_ an identity function.)

divyekapoor commented:
Hi Piotr,
Thanks for the explanation. The issue I'm raising is the interaction
between h1 and h2. h2 is only exposed to keys in a space that is the subset
of pathological cases for h1. What needs to be shown is that the output
distribution uniformity for h2 is maintained for this pathological subset
(ie. The overlap of pathological cases for h1 and h2 is either null or
bounded to a small value). This is possibly a stronger guarantee than just
cryptographic strength of the hash function. It may be the case that
SipHash meets this requirement through its construction (I do not have any
indications otherwise) but having a bound on the deviates from uniformity
would be great to avoid surprises. Regardless, thanks for all the work on
this. Appreciate it.

Cheers!
Divye
On Thu, Jul 14, 2016 at 12:12 PM Piotr Czarnecki notifications@github.com
wrote:

> @divyekapoor https://github.com/divyekapoor Adaptive Robin Hood is not
> a new hash function, it's a way of using two existing hash functions, h₁
> and h₂, at different times. The latter is a cryptographic hash function,
> whereas the former is just a hash function with good statistical
> properties*. This implementation will use SipHash-1-3 as the latter, and
> Farmhash as the former.
> 
> Still, the name "adaptive hashing" is misleading. I'd like to use one of
> the following short names.
> - Adaptive Robin Hood
> - Fused Robin Hood hashing
> - Safeguarded Robin Hood hashing
> - Firewalled Robin Hood hashing
> - All cryptographic hash functions have good avalanche effect, but not all
>   hash functions that have good avalanche effect are cryptographic.
> 
> @arthurprs https://github.com/arthurprs You're right about some
> languages using an identity function as h₁. However, HashMap in Rust has
> different needs, design choices and guarantees compared to many chaining
> hash maps. (Xoring against the seed doesn't help.)
> 
> Thanks for the benchmarks. Yes, they show oddly low speed for the
> finalizer hash. It does 2 multiplications and 6 basic operations, which
> should be very fast on most CPUs. (This is also an argument for _not
> using_ an identity function.)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/contain-rs/hashmap2/pull/5#issuecomment-232763403,
> or mute the thread
> https://github.com/notifications/unsubscribe/AANYcOuy_dFf6CJoWIY-LOU2WPtmADrAks5qVooTgaJpZM4H2XyN
> .

divyekapoor commented:
PS: I don't want to be a pain here. An empirical eval over a large keyspace
might be good enough. :)
On Thu, Jul 14, 2016 at 3:21 PM Divye Kapoor divyekapoor@gmail.com wrote:

> Hi Piotr,
> Thanks for the explanation. The issue I'm raising is the interaction
> between h1 and h2. h2 is only exposed to keys in a space that is the subset
> of pathological cases for h1. What needs to be shown is that the output
> distribution uniformity for h2 is maintained for this pathological subset
> (ie. The overlap of pathological cases for h1 and h2 is either null or
> bounded to a small value). This is possibly a stronger guarantee than just
> cryptographic strength of the hash function. It may be the case that
> SipHash meets this requirement through its construction (I do not have any
> indications otherwise) but having a bound on the deviates from uniformity
> would be great to avoid surprises. Regardless, thanks for all the work on
> this. Appreciate it.
> 
> Cheers!
> Divye
> On Thu, Jul 14, 2016 at 12:12 PM Piotr Czarnecki notifications@github.com
> wrote:
> 
> > @divyekapoor https://github.com/divyekapoor Adaptive Robin Hood is not
> > a new hash function, it's a way of using two existing hash functions, h₁
> > and h₂, at different times. The latter is a cryptographic hash function,
> > whereas the former is just a hash function with good statistical
> > properties*. This implementation will use SipHash-1-3 as the latter, and
> > Farmhash as the former.
> > 
> > Still, the name "adaptive hashing" is misleading. I'd like to use one of
> > the following short names.
> > - Adaptive Robin Hood
> > - Fused Robin Hood hashing
> > - Safeguarded Robin Hood hashing
> > - Firewalled Robin Hood hashing
> > - All cryptographic hash functions have good avalanche effect, but not
> >   all hash functions that have good avalanche effect are cryptographic.
> > 
> > @arthurprs https://github.com/arthurprs You're right about some
> > languages using an identity function as h₁. However, HashMap in Rust has
> > different needs, design choices and guarantees compared to many chaining
> > hash maps. (Xoring against the seed doesn't help.)
> > 
> > Thanks for the benchmarks. Yes, they show oddly low speed for the
> > finalizer hash. It does 2 multiplications and 6 basic operations, which
> > should be very fast on most CPUs. (This is also an argument for _not
> > using_ an identity function.)
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > https://github.com/contain-rs/hashmap2/pull/5#issuecomment-232763403,
> > or mute the thread
> > https://github.com/notifications/unsubscribe/AANYcOuy_dFf6CJoWIY-LOU2WPtmADrAks5qVooTgaJpZM4H2XyN
> > .

pczarn commented:
I see! I need to correct my explanation. Adaptive RH is a way of using two hash function families, H₁={h₁} and H₂, at different times. We specify that h₁ has good statistical properties, and H₂ is pairwise independent. This implementation will use h₁=FarmHash(seed=0) and H₂=SipHash.

Suppose for a moment that we choose h₁=SipHash(seed), for any constant seed. In other words, h₁∈H₂. In this worst-case situation, the probability a Rust thread\* getting pathological overlap with adaptive RH is 2⁻¹²⁸.

I think that given hash functions a, b and c, if {a, b} are independent and {a, c} are entirely dependent for a subset of their domain, then {b, c} are independent for that subset. It seems we don't need to worry about the overlap of pathological cases. I may be wrong and perhaps this is not enough proof.

\* Rust now generates a random key per system thread.

arthurprs commented:
@pczarn I ran some benches against sip13 and it even beats the adaptive one in some cases. Any idea how/why?

code is here: https://github.com/arthurprs/hashmap2/tree/finalizer

```
     Running target/release/bench_adaptive-0d973e825f6be877

running 10 tests
test find_existing     ... bench:      29,650 ns/iter (+/- 2,169)
test find_nonexisting  ... bench:      30,293 ns/iter (+/- 1,376)
test get_remove_insert ... bench:         140 ns/iter (+/- 18)
test grow_by_insertion ... bench:         113 ns/iter (+/- 12)
test hashmap_as_queue  ... bench:          78 ns/iter (+/- 5)
test new_drop          ... bench:           1 ns/iter (+/- 0)
test new_insert_drop   ... bench:          72 ns/iter (+/- 3)
test rev_insert        ... bench:       4,475 ns/iter (+/- 250)
test rng_insert        ... bench:       4,036 ns/iter (+/- 219)
test seq_insert        ... bench:       4,384 ns/iter (+/- 260)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured

     Running target/release/bench_with_siphash-3f3992eb47598559

running 10 tests
test find_existing     ... bench:      26,299 ns/iter (+/- 2,660)
test find_nonexisting  ... bench:      26,930 ns/iter (+/- 2,446)
test get_remove_insert ... bench:         134 ns/iter (+/- 7)
test grow_by_insertion ... bench:         120 ns/iter (+/- 20)
test hashmap_as_queue  ... bench:          77 ns/iter (+/- 8)
test new_drop          ... bench:         116 ns/iter (+/- 10)
test new_insert_drop   ... bench:         184 ns/iter (+/- 14)
test rev_insert        ... bench:       4,242 ns/iter (+/- 338)
test rng_insert        ... bench:       4,663 ns/iter (+/- 439)
test seq_insert        ... bench:       3,754 ns/iter (+/- 352)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured
```

pczarn commented:
I ran benches again.

#### Adaptive RH (Mixing Finalizer)

```
test find_existing     ... bench:      12,647 ns/iter (+/- 433)
test find_nonexisting  ... bench:      11,381 ns/iter (+/- 4,906)
test get_remove_insert ... bench:          91 ns/iter (+/- 40)
test grow_by_insertion ... bench:          99 ns/iter (+/- 76)
test hashmap_as_queue  ... bench:          85 ns/iter (+/- 26)
test new_drop          ... bench:           3 ns/iter (+/- 0)
test new_insert_drop   ... bench:          74 ns/iter (+/- 6)
```

#### SipHash-1-3

```
test find_existing     ... bench:      20,670 ns/iter (+/- 179)
test find_nonexisting  ... bench:      19,198 ns/iter (+/- 2,395)
test get_remove_insert ... bench:         128 ns/iter (+/- 5)
test grow_by_insertion ... bench:         167 ns/iter (+/- 46)
test hashmap_as_queue  ... bench:          77 ns/iter (+/- 3)
test new_drop          ... bench:          10 ns/iter (+/- 0)
test new_insert_drop   ... bench:          85 ns/iter (+/- 5)
```

@arthurprs I can't see why. Could you check if the faster path is really running, with `map.hasher().uses_safe_hashing()`?

pczarn committed:
Update for SipHash13

pczarn committed:
doc

pczarn committed:
inline things

arthurprs commented:
I used the latest version from your branch and the results make much more sense now, not sure why my branch was so broken. We can probably disregard the results I had for the identity hash.

```
     Running target/release/bench_adaptive-0d973e825f6be877

running 7 tests
test find_existing     ... bench:       8,879 ns/iter (+/- 597)
test find_nonexisting  ... bench:       7,207 ns/iter (+/- 634)
test get_remove_insert ... bench:          72 ns/iter (+/- 7)
test grow_by_insertion ... bench:         123 ns/iter (+/- 80)
test hashmap_as_queue  ... bench:          50 ns/iter (+/- 6)
test new_drop          ... bench:           1 ns/iter (+/- 1)
test new_insert_drop   ... bench:          57 ns/iter (+/- 29)

test result: ok. 0 passed; 0 failed; 0 ignored; 7 measured

     Running target/release/bench_with_siphash-3f3992eb47598559

running 7 tests
test find_existing     ... bench:      14,415 ns/iter (+/- 2,229)
test find_nonexisting  ... bench:      12,601 ns/iter (+/- 2,570)
test get_remove_insert ... bench:         101 ns/iter (+/- 15)
test grow_by_insertion ... bench:         111 ns/iter (+/- 100)
test hashmap_as_queue  ... bench:          61 ns/iter (+/- 14)
test new_drop          ... bench:           5 ns/iter (+/- 2)
test new_insert_drop   ... bench:          61 ns/iter (+/- 15)

test result: ok. 0 passed; 0 failed; 0 ignored; 7 measured

     Running target/release/hashmap2-2f26d306a0cc1a8d
```

divyekapoor commented:
@pczarn Proof looks good. The only thing left to establish is that the two hash functions (or hash families) are indeed independent (ie. correlation is approx. 0) which is likely the case.

Thank you.

Veedrac commented:
Why is this only for simple key types? String hashes could do with adaptive hashing too, no?

arthurprs commented:
@pczarn can you think of any way to minimize the code-complexity of this proposal?

bstrie commented:
@pczarn If the collision threshold is N, what's to stop an attacker from attempting to collide many different buckets N-1 times in order to DoS the server without triggering the collision detector?

ticki commented:
@bstrie How much different would that be from just spamming keys randomly to fill the hashtable up? Clearly, it is limited how much you can delay a particular key.

comex commented:
@ticki Very different, as the expected number of probes per lookup for random keys is constant - something like 3, AFAIK, depending on the load factor - no matter how big the hash table gets, while the worst case before hitting the limit is 127 probes.

If the u64 hash values are different (only equal modulo the table size), that's just 127 u64 comparisons, which is a bit slower but no big deal.  If they're the same, though, that could be 127 comparisons of really long strings.  In theory (as is mentioned in one of the todo comments) this could be a problem even with a _single_ chain that gets repeatedly looked up - lookups aren't much cheaper than insertions, so AFAICS there's no real need to blow up the table... 

This should be solvable by specifically checking for equal hashes.  Set a much lower chain length limit for the fast path, like 16 or so - which should still be very uncommon.  Every insertion that exceeds that chain length triggers a separate function that scans the chain to see if many keys (say, also 16) have equal hashes.  If this is true, perform the switch, and of course still do it if the chain length exceeds 128 even with unequal hashes.

If the hasher truly behaves randomly (as it really should for non-attack scenarios), the chance of even one 64-bit collision should be [rather low](https://en.wikipedia.org/wiki/Birthday_attack), and every additional collision required [divides the probability by on the order of 2^64](http://link.springer.com.sci-hub.cc/chapter/10.1007%2F11927587_5?LI=true).

Well, technically not just attack scenarios: it's possible to end up with a non-malicious but systemic source of full collisions, such as if someone's custom Hash implementation hashes only part of the object.  That usually means the _input_ to the hasher is the same, so with enough collisions the hash table is doomed to failure no matter what hash it uses (or if anything, switching to SipHash may be beneficial).  But if there are only a few objects with equal hasher input, my proposal would make a useless switch more likely.  Since this would already be a serious bug and SipHash is not _that_ slow, I don't think that's a big deal.

ticki commented:
@comex

That's not really my point. Say you want to attack key `K`, then you generate a bunch of preimages of `hash(K) % E`. You can maximally insert `N` of these before switching hash function. Due to probing, doing so to another key simultaneously would require that the entry is at minimum `N` entries away from any other attacked keys. This means that you can at most attack `E / N` entries, i.e. you slow down `E / N` keys. Well, you'd need `N` insertions for each attack, hence `E` insertions to complete such an attack. That's no different from just repeatedly inserting without any knowledge about the internals.

The only advantage over the uninformed attack is that you partially get to choose which keys to slow down, but even you can only slow it down by `N` probes.

@pczarn 

An idea: When reallocating the table, it should check the highest probe length and conditionally switch back to the fast hash function.

ticki commented:
I'm fine with having `N = 16`. A random hash function in a table of 1024 entries should emit such behavior naturally with a probability less than 0.0001% (from the collision counting equation).

Edit: My calculation was off.

Veedrac commented:
@comex 

> Set a much lower chain length limit for the fast path, like 16 or so - which should still be very uncommon.

Sadly strong chain length bounds only apply with the double-probing variant of Robin Hood hashing. Chains regularly get to length ~46 for million-long maps with the linear probing variant.

Insertions have much more worrisome behaviour, since they can end up shifting 1k elements even with purely random elements in the worst case.

ticki commented:
@Veedrac Yeah, clustering can have a real, negative effect. One possible solution is to use quadratic + Robin Hood. That should make it much less likely to happen.

comex commented:
@ticki Slowing down isn't all or nothing.  If you insert N keys with the same hash-modulo-table-size, the first takes 0 probes, the second takes 1, ..., the Nth takes N-1; total number of probes is N(N-1)/2.    That's for each chain; the current E doesn't really matter, since you can just repeat this an arbitrary number of times (as long as you know where in the table there's free space) and let the table be grown, ideally having the hash also be equal modulo the new table size - in which case each resize repeats all the work done so far, asymptotically doubling the total number of probes.  

The average number of probes per insertion (which we understand as a cost to the attacker) is thus roughly (N-1)/2 without growth - for N=128, that's 63.5 - while inserting keys randomly keeps it under 3.  Double both numbers to factor in growth.

Anyway, that's the worst case for the attacker, if (a) they can only perform an arbitrary number of insertions, not lookups, and (b) the inserted keys are required to be unique (e.g. the program refuses the request if an insertion finds an existing key).  If they're able to either keep looking up the same key or keep re-inserting it, they can just hammer the last key in the chain, N probes for each operation.  In this case the only incentive to create multiple chains is to pessimize cache behavior.

@Veedrac Hrm.  If my suggestion to have an intermediate step of checking for fully equal keys is followed, having a few chains >= 16 but < 128 is not the end of the world.  I guess it depends how cheap the check can be made...

But if an alternate probing scheme can avoid the code bloat of adding that logic while improving performance in general, it sounds like a good idea even at the expense of some code bloat of its own.  Pure double hashing has bad cache behavior, but what about starting with linear probing and switching to double hashing after some low iteration count (like 4)?  That is, the probe locations would be h1, h1+1, h1+2, h1+3, h1+h2, h1+2h2, etc.  Or maybe ..., h1+3, h1+h2, h1+h2+1, h1+h2+2, h2+h2+3, h2+2h2, ...

ticki commented:
@comex You forget the fact that collisions are not merely enough. You have to deal with probe length, i.e. you need to keep `N` distance from other hashes, which reduces the damage.

pczarn commented:
> Why is this only for simple key types?

The code doesn't cover strings yet. The algorithm will eventually work just fine for strings and slices.

> can you think of any way to minimize the code-complexity of this proposal?

I can't think of any such way, despite lots of consideration. This proposal is already reasonably simple. If someone can come up with a simplification, I would be impressed and grateful. I still have to write an RFC.

> This should be solvable by specifically checking for equal hashes. Set a much lower chain length limit for the fast path, like 16 or so - which should still be very uncommon. Every insertion that exceeds that chain length triggers a separate function that scans the chain to see if many keys (say, also 16) have equal hashes. If this is true, perform the switch, and of course still do it if the chain length exceeds 128 even with unequal hashes.

When a chain turns out not to have many equal hashes, you need to resume insertion. The hard part is implementing the resuming it in a way that won't harm code generation. A recursive call is not ideal for generated code, unless it gets tail call optimization.

Should I be more concerned about equal hashes? I think applications should at least ensure their inputs are not too long. Web servers certainly check the length of query keys.

Typical DDoS is devastating for many hashmaps, e.g. in Java, because every key comparison requires reading a heap-allocated object, which loads a cache line. With large strings, we're reading contiguous memory.

> An idea: When reallocating the table, it should check the highest probe length and conditionally switch back to the fast hash function.

I think it's not necessary. Why implement switching back, when the algorithm is meant to never switch in practice.

comex commented:
@ticki I don't know what you mean.  It should be possible to fill up the hash table without gaps, up to the maximum number of entries before resize - like this, imagining N were 4:

```
slot  0 1 2 3 4 5 6 7 8 9 a b
hash  0 0 0 0 4 4 4 4 8 8 8 8 ...
```

ticki commented:
@comex. Good point.

@pczarn 

> I think it's not necessary. Why implement switching back, when the algorithm is meant to never switch in practice.

Well, let's consider a strict hypothetical scenario:

You use the hashtable for a long running KV store (say, a server), and some attack generates a lot of collisions and inserts them into the hash table. Now, the hash table will move on to a secure hash function. When the attacker realize that their approach is inefficient, they might discontinue the attack. If the hash table keeps growing, and then reallocates, nothing is lost by switching back to the old hash function.

Of course, this gives a new attack vector: To do the collision attack, then let it reallocate another time, switching back to the old function, and repeat indefinite. Although, it is worth noting that getting a table to reallocate is not exactly trivial, since it requires you to insert a high (exponentially growing) number of insertions.

Generally, I think the best way of modeling the security in this case is to compare the attack to the naïve one in which a lot of random keys are spammed, given that it requires this technique to revert it back to an unsecure hash function, it is only marginally worse than having a "never-go-back" approach.

The question is this one: How big is the gain, and is it worth it? Unfortunately, that cannot be measured by micro benchmarks.

pczarn committed:
Fixes for nightly Rust

pczarn committed:
Small refactoring

pczarn committed:
Major refactor

pczarn committed:
Fix insertion under an incorrect hash

pczarn committed:
Refactor

pczarn committed:
try insertion and robin_hood instead of search

pczarn committed:
update

pczarn committed:
Add flag for displacement

pczarn committed:
refactor

pczarn committed:
Remove unused code

pczarn commented:
I changed the way the safeguard works. Previously, it allowed for huge cost of map creation, only disallowing costly lookups. Now, the safeguard is a part of insertion instead of every lookup.

The code is a bit less complex.

Two tasks.
* [x] found the equation for the probability that a randomly picked bucket is in a chain (island) of length X.
* [x] wrote an RFC.

arthurprs commented:
Very cool. It can be seen as an adaptive load factor now, which is great. Maybe it should take into account resizing policy though.

funny-falcon commented:
It is quite easy to add seed to this mix function:

https://en.wikipedia.org/wiki/Xor-encrypt-xor
> In 1991, motivated by Rivest's DESX construction, Even and Mansour proposed a much simpler scheme (the "two-key Even-Mansour scheme"), which they suggested was perhaps the simplest possible block cipher: XOR the plaintext with a prewhitening key, apply a publicly known unkeyed permutation (in practice, a pseudorandom permutation) to the result, and then XOR a postwhitening key to the permuted result to produce the final ciphertext.[3][4]

So if we have 128bit seed, then we can simply:
```rust
self.hash = mix(msg_data ^ seed.k0) ^ seed.k1;
```
This construction will be strong enough to not use fallback to SipHash at all (for simple keys).

(note: Even-Mansour scheme relies on "strong pseudorandom permutation".
`mix` function is pseudorandom permutation with unknown "strength".
But I'm pretty sure, it is "strong enough" for this use-case, ie as a hash function in a hash table).

funny-falcon commented:
> This construction will be strong enough to not use fallback to SipHash at all (for simple keys).

Then there is no need in Adaptive Hashing for simple keys.
But Adaptive Hashing still could be useful for "more complex" keys (ie strings, structs, etc.), so we just need fast hash function for them.
